src/test/java/ArrayIndexedBagTest.java

```diff
@@ -0,0 +1,258 @@
+        assertThrows(IndexOutOfBoundsException.class, () -> bag.get(99));
+    }
+
+    @Test
+    void getAndIndexOfCorrespondOnValueAndIndexInIndexedBag() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(101);
+        bag.add(102);
+        bag.add(103);
+        bag.add(104);
+        bag.add(2, 99);
+        assertEquals(bag.get(bag.indexOf(99)), 99);
+        assertEquals(bag.indexOf(bag.get(2)), 2);
+    }
+
+    @Test
+    void settingElementAtIndexOfIndexedBagUpdatesValue() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(110);
+        bag.set(1, 55);
+        assertEquals(55, bag.get(1));
+    }
+
+    @Test
+    void settingAnIndexNotInRangeOfIndexedBagThrowsException() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        assertThrows(IndexOutOfBoundsException.class, () -> bag.set(99, 101));
+    }
+
+    @Test
+    void removeByValueFromIndexedBagReturnsValue() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        assertEquals(101, bag.remove(Integer.valueOf(101)));
+    }
+
+    @Test
+    void removeByValueRemovesElementFromIndexedBag() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        bag.remove(Integer.valueOf(101));
+        assertFalse(bag.contains(101));
+    }
+
+    @Test
+    void removeByValueRemovesSingleElementFromIndexedBag() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(99);
+        bag.add(99);
+        bag.remove(Integer.valueOf(99));
+        assertEquals(2, bag.getCount(99));
+    }
+
+    @Test
+    void removeByValueOfIndexedBagUpdatesSize() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        bag.remove(Integer.valueOf(101));
+        assertEquals(2, bag.size());
+    }
+
+    @Test
+    void removeByValueOfNonExistingElementInIndexedBagThrowsException() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        assertThrows(NoSuchElementException.class, () -> bag.remove(Integer.valueOf(110)));
+    }
+
+    @Test
+    void removeByIndexFromIndexedBagReturnsValue() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        assertEquals(101, bag.remove(1));
+    }
+
+    @Test
+    void removeByIndexRemovesElementFromIndexedBag() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        bag.remove(1);
+        assertFalse(bag.contains(101));
+    }
+
+    @Test
+    void removeByIndexOfIndexedBagUpdatesSize() {
```

Comment:
May also want to add a test which removes the sole element from a bag, causing it to be `isEmpty`.